File: python/src/bleepstore/storage/azure.py
"""Azure Blob Storage gateway backend for BleepStore.

Proxies all data operations to an upstream Azure Blob Storage container
via azure-storage-blob (async). Metadata stays in local SQLite — this
backend handles raw bytes only.

Key mapping:
    Objects:  {prefix}{bleepstore_bucket}/{key}

Multipart strategy uses Azure Block Blob primitives:
    put_part()       → stage_block() on the final blob (no temp objects)
    assemble_parts() → commit_block_list() to finalize
    delete_parts()   → no-op (uncommitted blocks auto-expire in 7 days)

Credentials are resolved via DefaultAzureCredential (env vars, managed
identity, Azure CLI, etc.).
"""

import base64
import hashlib
import logging
from collections.abc import AsyncIterator

from azure.core.exceptions import ResourceNotFoundError
from azure.identity.aio import DefaultAzureCredential
from azure.storage.blob.aio import ContainerClient
from azure.storage.blob import BlobBlock

logger = logging.getLogger(__name__)

# Streaming chunk size: 64 KB (matches local, AWS, GCP backends)
_CHUNK_SIZE = 64 * 1024


class AzureGatewayBackend:
    """Storage backend that proxies to an Azure Blob Storage container.

    All BleepStore buckets/objects are stored under a single upstream
    Azure container with a key prefix to namespace them.

    Attributes:
        container_name: The upstream Azure Blob container name.
        account_url: The Azure Storage account URL (e.g. https://account.blob.core.windows.net).
        prefix: Key prefix for all blobs in the upstream container.
    """

    def __init__(
        self,
        container_name: str,
        account_url: str = "",
        prefix: str = "",
    ) -> None:
        self.container_name = container_name
        self.account_url = account_url
        self.prefix = prefix
        self._container_client: ContainerClient | None = None
        self._credential: DefaultAzureCredential | None = None

    def _blob_name(self, bucket: str, key: str) -> str:
        """Map a BleepStore bucket/key to an upstream Azure blob name."""
        return f"{self.prefix}{bucket}/{key}"

    @staticmethod
    def _block_id(upload_id: str, part_number: int) -> str:
        """Generate a block ID for Azure staged blocks.

        Block IDs must be base64-encoded and the same length for all
        blocks in a blob. Includes upload_id to avoid collisions between
        concurrent multipart uploads to the same key.
        """
        return base64.b64encode(f"{upload_id}:{part_number:05d}".encode()).decode()

    async def init(self) -> None:
        """Create the Azure ContainerClient and verify the container exists.

        Raises:
            ValueError: If the upstream container does not exist or is inaccessible.
        """
        self._credential = DefaultAzureCredential()
        self._container_client = ContainerClient(
            self.account_url,
            self.container_name,
            credential=self._credential,
        )

        # Verify container exists
        try:
            exists = await self._container_client.exists()
        except Exception as e:
            await self._container_client.close()
            await self._credential.close()
            self._container_client = None
            self._credential = None
            raise ValueError(
                f"Cannot access upstream Azure container '{self.container_name}': {e}"
            ) from e

        if not exists:
            await self._container_client.close()
            await self._credential.close()
            self._container_client = None
            self._credential = None
            raise ValueError(f"Upstream Azure container '{self.container_name}' does not exist")

        logger.info(
            "Azure gateway backend initialized: container=%s account=%s prefix='%s'",
            self.container_name,
            self.account_url,
            self.prefix,
        )

    async def close(self) -> None:
        """Close the Azure client session."""
        if self._container_client is not None:
            await self._container_client.close()
            self._container_client = None
        if self._credential is not None:
            await self._credential.close()
            self._credential = None

    async def put(self, bucket: str, key: str, data: bytes) -> str:
        """Upload an object to the upstream Azure container.

        Computes MD5 locally for a consistent ETag.

        Returns:
            The hex-encoded MD5 of the stored data.
        """
        blob_name = self._blob_name(bucket, key)
        md5 = hashlib.md5(data).hexdigest()

        blob_client = self._container_client.get_blob_client(blob_name)
        await blob_client.upload_blob(data, overwrite=True)
        return md5

    async def put_stream(
        self,
        bucket: str,
        key: str,
        stream: AsyncIterator[bytes],
        content_length: int | None = None,
    ) -> tuple[str, int]:
        """Stream-write: collect stream and delegate to put()."""
        chunks = []
        async for chunk in stream:
            chunks.append(chunk)
        data = b"".join(chunks)
        md5_hex = await self.put(bucket, key, data)
        return md5_hex, len(data)

    async def put_part_stream(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        part_number: int,
        stream: AsyncIterator[bytes],
        content_length: int | None = None,
    ) -> tuple[str, int]:
        """Stream-write part: collect stream and delegate to put_part()."""
        chunks = []
        async for chunk in stream:
            chunks.append(chunk)
        data = b"".join(chunks)
        md5_hex = await self.put_part(bucket, key, upload_id, part_number, data)
        return md5_hex, len(data)

    async def get(self, bucket: str, key: str) -> bytes:
        """Download an object from the upstream Azure container.

        Raises:
            FileNotFoundError: If the object does not exist.
        """
        blob_name = self._blob_name(bucket, key)
        blob_client = self._container_client.get_blob_client(blob_name)

        try:
            downloader = await blob_client.download_blob()
            return await downloader.readall()
        except ResourceNotFoundError as e:
            raise FileNotFoundError(f"Object not found: {bucket}/{key}") from e

    async def get_stream(
        self, bucket: str, key: str, offset: int = 0, length: int | None = None
    ) -> AsyncIterator[bytes]:
        """Stream an object from the upstream Azure container in 64KB chunks.

        Raises:
            FileNotFoundError: If the object does not exist.
        """
        blob_name = self._blob_name(bucket, key)
        blob_client = self._container_client.get_blob_client(blob_name)

        kwargs: dict = {}
        if offset > 0:
            kwargs["offset"] = offset
        if length is not None:
            kwargs["length"] = length

        try:
            downloader = await blob_client.download_blob(**kwargs)
        except ResourceNotFoundError as e:
            raise FileNotFoundError(f"Object not found: {bucket}/{key}") from e

        async for chunk in downloader.chunks():
            yield chunk

    async def delete(self, bucket: str, key: str) -> None:
        """Delete an object from the upstream Azure container.

        Idempotent — catches ResourceNotFoundError silently.
        """
        blob_name = self._blob_name(bucket, key)
        blob_client = self._container_client.get_blob_client(blob_name)

        try:
            await blob_client.delete_blob()
        except ResourceNotFoundError:
            pass  # Idempotent: treat as success

    async def exists(self, bucket: str, key: str) -> bool:
        """Check if an object exists in the upstream Azure container."""
        blob_name = self._blob_name(bucket, key)
        blob_client = self._container_client.get_blob_client(blob_name)
        return await blob_client.exists()

    async def put_part(
        self, bucket: str, key: str, upload_id: str, part_number: int, data: bytes
    ) -> str:
        """Stage a block on the final blob (Azure Block Blob multipart).

        Unlike AWS/GCP, parts are staged directly on the final blob using
        stage_block(). No temporary objects are created. Uncommitted blocks
        auto-expire in 7 days.

        Returns:
            The hex-encoded MD5 of the part data.
        """
        blob_name = self._blob_name(bucket, key)
        blob_client = self._container_client.get_blob_client(blob_name)
        block_id = self._block_id(upload_id, part_number)
        md5 = hashlib.md5(data).hexdigest()

        await blob_client.stage_block(block_id, data, length=len(data))
        return md5

    async def assemble_parts(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        part_numbers: list[int],
    ) -> str:
        """Commit staged blocks into the final blob.

        Builds a block list from the upload_id and part numbers, then
        calls commit_block_list() to finalize the blob. Downloads the
        result to compute a consistent MD5 ETag.

        Returns:
            The hex-encoded MD5 of the assembled object.
        """
        blob_name = self._blob_name(bucket, key)
        blob_client = self._container_client.get_blob_client(blob_name)

        block_list = [BlobBlock(block_id=self._block_id(upload_id, pn)) for pn in part_numbers]
        await blob_client.commit_block_list(block_list)

        # Download the committed blob to compute MD5
        downloader = await blob_client.download_blob()
        data = await downloader.readall()
        return hashlib.md5(data).hexdigest()

    async def delete_parts(self, bucket: str, key: str, upload_id: str) -> None:
        """No-op — uncommitted Azure blocks auto-expire in 7 days.

        Unlike AWS/GCP, there are no temporary part objects to clean up.
        Azure automatically garbage-collects uncommitted blocks.
        """
        pass

    async def copy_object(
        self,
        src_bucket: str,
        src_key: str,
        dst_bucket: str,
        dst_key: str,
    ) -> str:
        """Copy an object using Azure server-side copy.

        Builds the source URL from account URL, container, and blob name,
        then uses start_copy_from_url() for a server-side copy. Downloads
        the destination to compute a consistent MD5 ETag.

        Returns:
            The hex-encoded MD5 ETag of the copied object.
        """
        src_blob_name = self._blob_name(src_bucket, src_key)
        dst_blob_name = self._blob_name(dst_bucket, dst_key)

        # Build source URL
        source_url = f"{self.account_url}/{self.container_name}/{src_blob_name}"

        dst_blob_client = self._container_client.get_blob_client(dst_blob_name)
        await dst_blob_client.start_copy_from_url(source_url)

        # Download destination to compute MD5
        downloader = await dst_blob_client.download_blob()
        data = await downloader.readall()
        return hashlib.md5(data).hexdigest()

File: golang/internal/storage/azure.go
// Package storage provides the Azure Blob Storage gateway backend for BleepStore.
//
// The Azure gateway backend proxies all data operations to an upstream Azure
// Blob Storage container via the official Azure SDK for Go. Metadata stays
// in local SQLite -- this backend handles raw bytes only.
//
// Key mapping:
//
//	Objects:  {prefix}{bleepstore_bucket}/{key}
//
// Multipart strategy uses Azure Block Blob primitives:
//
//	put_part()       → StageBlock() on the final blob (no temp objects)
//	assemble_parts() → CommitBlockList() to finalize
//	delete_parts()   → no-op (uncommitted blocks auto-expire in 7 days)
//
// Credentials are resolved via DefaultAzureCredential (env vars, managed
// identity, Azure CLI, etc.).
package storage

import (
	"bytes"
	"context"
	"crypto/md5"
	"encoding/base64"
	"fmt"
	"io"
	"log/slog"
	"strings"
)

// AzureBlobAPI defines the subset of the Azure Blob Storage client interface
// that the gateway backend uses. This allows mocking in tests.
type AzureBlobAPI interface {
	// UploadBlob uploads data to a blob, overwriting if it already exists.
	UploadBlob(ctx context.Context, containerName, blobName string, data []byte) error
	// DownloadBlob downloads a blob's contents.
	DownloadBlob(ctx context.Context, containerName, blobName string) ([]byte, error)
	// DeleteBlob deletes a blob. Returns an error if the blob does not exist.
	DeleteBlob(ctx context.Context, containerName, blobName string) error
	// BlobExists checks if a blob exists.
	BlobExists(ctx context.Context, containerName, blobName string) (bool, error)
	// GetBlobProperties retrieves the size of a blob.
	GetBlobProperties(ctx context.Context, containerName, blobName string) (int64, error)
	// StartCopyFromURL copies a blob from a source URL.
	StartCopyFromURL(ctx context.Context, containerName, blobName, sourceURL string) error
	// StageBlock stages a block on a blob for later commit.
	StageBlock(ctx context.Context, containerName, blobName, blockID string, data []byte) error
	// CommitBlockList commits a list of block IDs to finalize a blob.
	CommitBlockList(ctx context.Context, containerName, blobName string, blockIDs []string) error
}

// AzureGatewayBackend implements the StorageBackend interface by proxying
// storage operations to Azure Blob Storage. This allows BleepStore to act
// as an S3-compatible gateway in front of Azure Blob.
//
// All BleepStore buckets/objects are stored under a single upstream Azure
// container with a key prefix to namespace them.
type AzureGatewayBackend struct {
	// Container is the upstream Azure Blob container name.
	Container string
	// AccountURL is the Azure storage account URL (e.g. https://account.blob.core.windows.net).
	AccountURL string
	// Prefix is the key prefix for all blobs in the upstream container.
	Prefix string
	// client is the Azure Blob client (satisfying AzureBlobAPI interface).
	client AzureBlobAPI
}

// NewAzureGatewayBackend creates a new AzureGatewayBackend configured to proxy
// to the specified Azure Blob container. It initializes the Azure SDK client
// using DefaultAzureCredential.
func NewAzureGatewayBackend(ctx context.Context, container, accountURL, prefix string) (*AzureGatewayBackend, error) {
	client, err := newRealAzureClient(accountURL)
	if err != nil {
		return nil, fmt.Errorf("creating Azure client: %w", err)
	}

	b := &AzureGatewayBackend{
		Container:  container,
		AccountURL: accountURL,
		Prefix:     prefix,
		client:     client,
	}

	// Verify the upstream container is accessible by checking if a non-existent blob exists.
	_, err = b.client.BlobExists(ctx, container, "\x00nonexistent\x00")
	if err != nil {
		return nil, fmt.Errorf("cannot access upstream Azure container %q: %w", container, err)
	}

	slog.Info("Azure gateway backend initialized", "container", container, "account", accountURL, "prefix", prefix)
	return b, nil
}

// NewAzureGatewayBackendWithClient creates an AzureGatewayBackend with a
// pre-configured Azure client. This is primarily used for testing with mock
// clients.
func NewAzureGatewayBackendWithClient(container, accountURL, prefix string, client AzureBlobAPI) *AzureGatewayBackend {
	return &AzureGatewayBackend{
		Container:  container,
		AccountURL: accountURL,
		Prefix:     prefix,
		client:     client,
	}
}

// blobName maps a BleepStore bucket/key to an upstream Azure blob name.
func (b *AzureGatewayBackend) blobName(bucket, key string) string {
	return b.Prefix + bucket + "/" + key
}

// blockID generates a block ID for Azure staged blocks.
// Block IDs must be base64-encoded and the same length for all blocks
// in a blob. Includes uploadID to avoid collisions between concurrent
// multipart uploads to the same key.
func blockID(uploadID string, partNumber int) string {
	return base64.StdEncoding.EncodeToString(
		[]byte(fmt.Sprintf("%s:%05d", uploadID, partNumber)),
	)
}

// PutObject uploads object data to the upstream Azure Blob container. It reads
// all data, computes MD5 locally for a consistent ETag, then uploads to Azure.
func (b *AzureGatewayBackend) PutObject(ctx context.Context, bucket, key string, reader io.Reader, size int64) (int64, string, error) {
	blobKey := b.blobName(bucket, key)

	// Read all data to compute MD5 locally. Azure may return different ETags,
	// so we compute our own for consistency.
	data, err := io.ReadAll(reader)
	if err != nil {
		return 0, "", fmt.Errorf("reading object data: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))

	if err := b.client.UploadBlob(ctx, b.Container, blobKey, data); err != nil {
		return 0, "", fmt.Errorf("uploading to Azure Blob: %w", err)
	}

	return int64(len(data)), etag, nil
}

// GetObject retrieves object data from the upstream Azure Blob container.
// Returns the data stream, the object size, and an empty ETag (metadata store
// holds the authoritative ETag). The caller is responsible for closing the
// returned ReadCloser.
func (b *AzureGatewayBackend) GetObject(ctx context.Context, bucket, key string) (io.ReadCloser, int64, string, error) {
	blobKey := b.blobName(bucket, key)

	// Get properties first for size.
	blobSize, err := b.client.GetBlobProperties(ctx, b.Container, blobKey)
	if err != nil {
		if isAzureNotFound(err) {
			return nil, 0, "", fmt.Errorf("object not found: %s/%s", bucket, key)
		}
		return nil, 0, "", fmt.Errorf("getting blob properties from Azure: %w", err)
	}

	data, err := b.client.DownloadBlob(ctx, b.Container, blobKey)
	if err != nil {
		if isAzureNotFound(err) {
			return nil, 0, "", fmt.Errorf("object not found: %s/%s", bucket, key)
		}
		return nil, 0, "", fmt.Errorf("getting object from Azure Blob: %w", err)
	}

	return io.NopCloser(bytes.NewReader(data)), blobSize, "", nil
}

// DeleteObject removes an object from the upstream Azure Blob container.
// Idempotent: catches not-found silently.
func (b *AzureGatewayBackend) DeleteObject(ctx context.Context, bucket, key string) error {
	blobKey := b.blobName(bucket, key)

	err := b.client.DeleteBlob(ctx, b.Container, blobKey)
	if err != nil {
		if isAzureNotFound(err) {
			return nil // Idempotent: treat as success
		}
		return fmt.Errorf("deleting object from Azure Blob: %w", err)
	}
	return nil
}

// CopyObject copies an object within the upstream Azure Blob container using
// Azure server-side copy. Downloads the result to compute MD5 for a consistent
// ETag.
func (b *AzureGatewayBackend) CopyObject(ctx context.Context, srcBucket, srcKey, dstBucket, dstKey string) (string, error) {
	srcBlobName := b.blobName(srcBucket, srcKey)
	dstBlobName := b.blobName(dstBucket, dstKey)

	// Build source URL.
	sourceURL := fmt.Sprintf("%s/%s/%s", b.AccountURL, b.Container, srcBlobName)

	err := b.client.StartCopyFromURL(ctx, b.Container, dstBlobName, sourceURL)
	if err != nil {
		if isAzureNotFound(err) {
			return "", fmt.Errorf("source object not found: %s/%s", srcBucket, srcKey)
		}
		return "", fmt.Errorf("copying object in Azure Blob: %w", err)
	}

	// Download destination to compute MD5 for consistent ETag.
	data, err := b.client.DownloadBlob(ctx, b.Container, dstBlobName)
	if err != nil {
		return "", fmt.Errorf("reading copied object for ETag: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))
	return etag, nil
}

// PutPart stages a block on the final blob (Azure Block Blob multipart).
// Unlike AWS/GCP, parts are staged directly on the final blob using
// StageBlock(). No temporary objects are created. Uncommitted blocks
// auto-expire in 7 days.
//
// Computes MD5 locally for a consistent ETag.
func (b *AzureGatewayBackend) PutPart(ctx context.Context, bucket, key, uploadID string, partNumber int, reader io.Reader, size int64) (string, error) {
	blobKey := b.blobName(bucket, key)
	blkID := blockID(uploadID, partNumber)

	// Read all data to compute MD5 locally.
	data, err := io.ReadAll(reader)
	if err != nil {
		return "", fmt.Errorf("reading part data: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))

	if err := b.client.StageBlock(ctx, b.Container, blobKey, blkID, data); err != nil {
		return "", fmt.Errorf("staging block in Azure Blob: %w", err)
	}

	return etag, nil
}

// AssembleParts commits staged blocks into the final blob. Builds a block list
// from the upload_id and part numbers, then calls CommitBlockList() to finalize.
// Downloads the result to compute a consistent MD5 ETag.
func (b *AzureGatewayBackend) AssembleParts(ctx context.Context, bucket, key, uploadID string, partNumbers []int) (string, error) {
	blobKey := b.blobName(bucket, key)

	blockIDs := make([]string, len(partNumbers))
	for i, pn := range partNumbers {
		blockIDs[i] = blockID(uploadID, pn)
	}

	if err := b.client.CommitBlockList(ctx, b.Container, blobKey, blockIDs); err != nil {
		return "", fmt.Errorf("committing block list in Azure Blob: %w", err)
	}

	// Download the committed blob to compute MD5.
	data, err := b.client.DownloadBlob(ctx, b.Container, blobKey)
	if err != nil {
		return "", fmt.Errorf("reading assembled object for ETag: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))
	return etag, nil
}

// DeleteParts is a no-op for Azure. Uncommitted Azure blocks auto-expire
// in 7 days. Unlike AWS/GCP, there are no temporary part objects to clean up.
func (b *AzureGatewayBackend) DeleteParts(ctx context.Context, bucket, key, uploadID string) error {
	// No-op: Azure automatically garbage-collects uncommitted blocks.
	return nil
}

// CreateBucket is a no-op for the Azure gateway backend. All BleepStore
// buckets share a single upstream Azure container with key prefixes, so
// there is nothing to create on the Azure side.
func (b *AzureGatewayBackend) CreateBucket(ctx context.Context, bucket string) error {
	return nil
}

// DeleteBucket is a no-op for the Azure gateway backend. Bucket data is scoped
// by key prefix, so there is nothing to delete on the Azure side. The metadata
// store handles the actual bucket record deletion.
func (b *AzureGatewayBackend) DeleteBucket(ctx context.Context, bucket string) error {
	return nil
}

// ObjectExists checks whether an object exists in the upstream Azure Blob container.
func (b *AzureGatewayBackend) ObjectExists(ctx context.Context, bucket, key string) (bool, error) {
	blobKey := b.blobName(bucket, key)

	exists, err := b.client.BlobExists(ctx, b.Container, blobKey)
	if err != nil {
		return false, fmt.Errorf("checking object existence in Azure Blob: %w", err)
	}
	return exists, nil
}

// HealthCheck verifies that the upstream Azure Blob container is accessible.
func (b *AzureGatewayBackend) HealthCheck(ctx context.Context) error {
	_, err := b.client.BlobExists(ctx, b.Container, "\x00nonexistent\x00")
	return err
}

// isAzureNotFound checks if an Azure error is a not-found error.
func isAzureNotFound(err error) bool {
	if err == nil {
		return false
	}
	msg := strings.ToLower(err.Error())
	if strings.Contains(msg, "not found") || strings.Contains(msg, "404") ||
		strings.Contains(msg, "blobnotfound") || strings.Contains(msg, "containernotfound") ||
		strings.Contains(msg, "the specified blob does not exist") ||
		strings.Contains(msg, "the specified container does not exist") {
		return true
	}
	return false
}

// Ensure AzureGatewayBackend implements StorageBackend at compile time.
var _ StorageBackend = (*AzureGatewayBackend)(nil)
